internal/core/src/common/Array.h

```diff
@@ -0,0 +1,457 @@
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<int64_t*>(data_)[i] =
+                        field_data.long_data().data(i);
+                    offsets_.push_back(sizeof(int64_t) * i);
+                }
+                break;
+            }
+            case ScalarArray::kFloatData: {
+                element_type_ = DataType::FLOAT;
+                length_ = field_data.float_data().data().size();
+                size_ = length_ * sizeof(float);
+                data_ = new char[size_];
+                offsets_.reserve(length_);
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<float*>(data_)[i] =
+                        field_data.float_data().data(i);
+                    offsets_.push_back(sizeof(float) * i);
+                }
+                break;
+            }
+            case ScalarArray::kDoubleData: {
+                element_type_ = DataType::DOUBLE;
+                length_ = field_data.double_data().data().size();
+                size_ = length_ * sizeof(double);
+                data_ = new char[size_];
+                offsets_.reserve(length_);
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<double*>(data_)[i] =
+                        field_data.double_data().data(i);
+                    offsets_.push_back(sizeof(double) * i);
+                }
+                break;
+            }
+            case ScalarArray::kStringData: {
+                element_type_ = DataType::STRING;
+                length_ = field_data.string_data().data().size();
+                offsets_.reserve(length_);
+                for (int i = 0; i < length_; ++i) {
+                    offsets_.push_back(size_);
+                    size_ += field_data.string_data().data(i).size();
+                }
+
+                data_ = new char[size_];
+                for (int i = 0; i < length_; ++i) {
+                    std::copy_n(field_data.string_data().data(i).data(),
+                                field_data.string_data().data(i).size(),
+                                data_ + offsets_[i]);
+                }
+                break;
+            }
+            default: {
+                //                PanicInfo("unsupported element type for array");
+            }
+        }
+    }
+
+    Array(char* data,
+          size_t size,
+          DataType element_type,
+          std::vector<uint64_t>&& element_offsets)
+        : data_(data),
+          size_(size),
+          element_type_(element_type),
+          offsets_(std::move(element_offsets)),
+          length_(element_offsets.size()) {
+    }
+
+    bool
+    operator==(const Array& arr) const {
+        if (element_type_ != arr.element_type_) {
+            return false;
+        }
+        if (length_ != arr.length_) {
+            return false;
+        }
+        for (int i = 0; i < length_; ++i) {
+            switch (element_type_) {
```

Comment:
put  switch out of for is better for performance, just like
switch (element_type_) {
   case DataType::INT64:
     for (int i = 0; i< length_; ++i) {
        get_data<int64_t>(i) != arr.get_data<int64_t>(i);
     }
  .....
};

better, can use a template function to replace travel:
template<typename T>
bool TravelDataEqual(const Array& arr) {
   for (int i = 0; i< length_; ++i) {
          if( get_data<T>(i) != arr.get_data<T>(i)) {
               return false;
          }
   }
   return true;
}

Reply:
got it